docs/proposals/component-registration.md

```diff
@@ -0,0 +1,320 @@
+
+
+## Use Cases
+
+This proposal primarily covers features that can be achieved internally by Kubernetes itself, but many of these features and changes are motivated by external use cases, which need enablement. Not all of these use cases will be completely enabled by this proposal.
+
+
+### Deployment
+
+As a k8s operator, I want a to be able to add/remove new components (instances or types) to a new or running cluster (ideally without manually registering it with the apiserver).
+
+As a k8s operator, I want a kubectl command that allows me to validate that the cluster has been deployed and is fully functional.
+
+As a k8s operator, I want a kubectl command that allows me to check which components are deployed/installed on a specific cluster in order to infer which features it supports.
+
+Note: Inferring features from components isn't very useful with vanilla k8s, but becomes more useful as 3rd party or optional components are added (e.g. OpenShift) and/or if components-controller is split into multiple components.
+
+### Debugging
+
+As a k8s operator, I want a single endpoint (and kubectl command) where I can see the current condition of the cluster so that I can debug observed failures (dropped connections, unresponsiveness, etc.).
+
+Note: The endpoint should contain cause where possible, but at least the current condition of the components.
+
+### Notification
+
+As a k8s operator, I want to be notified when the cluster requires human intervention so that I can intervene.
+
+As an external notification system (e.g. PagerDuty), I want an endpoint that responds to regular requests with the current cluster condition so that I can determine (with my own rule set) when to notify the k8s operator.
+
+Note: What condition requires human intervention is not a rule set that can be hard-coded, is likely to be different for each cloud platform, changes based on component configuration, and is likely to require a time-series database.
+
+### Reporting
+
+As a k8s operator, I want to see a report/graph of the readiness/uptime/availability of the cluster over time so that I can make/satisfy SLAs.
+
+Note: At least requires a time-series database and probably a reporting engine.
+
+### Self-Healing
+
+As an k8s operator, I want the cluster to self-heal when individual components crash so that I don't get notified or need to intervene.
+
+As a local process monitoring tool (e.g. monit), I want to be able to determine the condition of the component process I am monitoring to determine if it needs to be restarted.
+
+As a remote container/vm monitoring tool (e.g. bosh monitor), I want to be able to determine the health of the component container/vm I am monitoring.
+
+Note: Self-healing requires knowing how to deploy the cluster components, which is different for each cloud platform and changes based on component configuration.
+
```

Comment:
Notes:
- I've been using "k8s operator" to indicate someone who manages a k8s cluster, analogous to your "cluster administrator".
- "Client" is a little ambiguous. Do you mean someone with read/write who can create pods/services or someone with just read who pulls resource spec/status (perhaps to manage an external system, like a load balancer)? Do we need names for both?
- I don't want to use "ingress" here, to avoid confusion with the discussion happening in #13947. But there is some overlap in terms of enabling various load balancing methods. I don't want to have to solve all the problems they're trying to solve regarding complex external load balancers just to enable component registration.
- I think all the service discovery k8s user use cases are already covered, but we can specify them for clarity.

How about this:

### Discovery

As a k8s user, I want to get a service address and named port(s) that can be used to make internal-to-internal service requests (client-side load balancer). [existing functionality - kube-proxy]

As a k8s user, I want to get a service address and named port(s) that can be used to make internal-to-external service requests (server-side load balancing). [existing functionality - ServiceSpec.ExternalIPs; cloud-provider-specific]

As a k8s user, I want to get a list of *ready* service provider addresses and named ports that can be used to make internal-to-external service requests (client-side load balancing). [existing functionality - Endpoints.Subsets.Addresses]

As a k8s user, I want to get a list of *ready and not ready* service provider addresses and named ports that can be used to make external-to-internal service requests (reverse-proxy or client-side load balancer). [existing functionality - Endpoints.Subsets.Addresses & NotReadyAddresses]

As a k8s operator, I want to specify a complete (regardless of readiness) list of external service provider addresses and named ports that can be used to make internal-to-external service requests (ExternalServiceTargetSpec). I want k8s to handle probing these for readiness and caching their condition for users to query (ExternalServiceTargetStatus.Conditions & Endpoints.Subsets.Addresses).